[FEAT] 콕 찌르기 도구 제작 뷰 구현

teethemoji opened the pull request:
## ✅ 관련 이슈

Fork Tool 제작 뷰 구현 (#7)

 ## ✅ 구현/변경 사항 및 이유
- SceneKit을 사용해 포크, 스푼, 거품기, 뒤집개를 3D 오브젝트로 제작하고 화면 상단의 Scene 안에 불러옵니다.
- 탭한 종류 버튼에 따라서 화면에 보여지는 오브젝트를 교체합니다.
- 탭한 색상 버튼에 따라서 화면에 보여지는 오브젝트의 색상을 교체합니다.
- 탭한 버튼에는 까만 테두리를 둘러줍니다.

 ## ✅ PR Point

러프하게 작업하느라 코드가 많이 지저분합니다.. 300줄이 넘어가버려서 코드를 좀 분리하고 싶은데 어떤 방법이 좋을까요?

 ## ✅ 참고 사항

![포크제작뷰구현](https://user-images.githubusercontent.com/60784937/180901610-f6a6e8f7-9984-4ec6-8766-10f3a0ce4dcb.gif)

teethemoji committed:
[FEAT] 찔러보기 도구 제작 뷰 구현

teethemoji committed:
[HOTFIX] 어제 PR(#20)과 함께 들어갔던 내용들을 다시 올리기

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+    // MARK: stack of image buttons View 선언
+    private let styleButtonsStackView: UIStackView = {
+        
+        // 구분선을 만듭니다.
+        let divider = UIView()
+        divider.backgroundColor = UIColor.customLightGray
+        divider.translatesAutoresizingMaskIntoConstraints = false
+        
+        // 그룹명을 표시해주는 라벨을 만듭니다.
+        let label = UILabel()
+        label.text = " 종류"
+        label.textColor = .black
+        label.textAlignment = .left
+        label.font = .systemFont(ofSize: 24, weight: .semibold)
+        label.translatesAutoresizingMaskIntoConstraints = false
+        
+        // 스택 뷰를 만들고 그 안에 버튼들을 모아둡니다.
+        let buttonsStackView = UIStackView(arrangedSubviews: [])
+        buttonsStackView.distribution = .fillEqually
+        buttonsStackView.spacing = 16
```
이 부분들은 밖에서 따로 선언해주고, arrangedSubView에서 불러오면 가독성이 더 좋아질 것 같습니다

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+            
+            button.translatesAutoresizingMaskIntoConstraints = false
+            
+            // button의 사이즈를 정해줍니다.
+            NSLayoutConstraint.activate([
+                button.widthAnchor.constraint(equalToConstant: 101),
+                button.heightAnchor.constraint(equalToConstant: 138)
+            ])
+            
```
화면 비율에 따라 설정해주는게 좋지 않을까요?

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        // StackView 속 label뷰가 차지하는 비율, buttonStackView가 차지하는 비율을 정해줍니다.
+        NSLayoutConstraint.activate([
+            
+            // divider
+            divider.topAnchor.constraint(equalTo: stackView.topAnchor),
+            divider.heightAnchor.constraint(equalToConstant: 2),
+            divider.widthAnchor.constraint(equalTo: stackView.widthAnchor),
+            
+            // label
+            label.topAnchor.constraint(equalTo: divider.bottomAnchor, constant: 10),
+            label.heightAnchor.constraint(equalToConstant: 24),
+            label.widthAnchor.constraint(equalTo: stackView.widthAnchor),
+            
+            // buttonStackView
+            buttonsStackView.topAnchor.constraint(equalTo: label.bottomAnchor, constant: 18),
+            buttonsStackView.heightAnchor.constraint(equalTo: stackView.heightAnchor, multiplier: 0.6),
+            buttonsStackView.widthAnchor.constraint(equalTo: stackView.widthAnchor)
```
이 부분은 개행만으로도 어떤 부분에 대한 constraint인지 확인할 수 있으므로 주석은 없어도 될 것 같습니당

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        
+    }()
+    
+    // MARK: viewDidLoad()
```
함수명 자체를 굳이 마크주석으로 달아줄 필요는 없어보입니다!

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+    
+    private func addConstraints() {
+        
+        var constraints = [NSLayoutConstraint]()
+        
+        // MARK: forkCustomView의 Autolayout Constraints
+        constraints.append(forkCustomView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor))
+        constraints.append(forkCustomView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 12))
+        constraints.append(forkCustomView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -12))
+        constraints.append(forkCustomView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor))
+        
+        // MARK: sceneView의 Autolayout Constraints
+        
+        constraints.append(sceneView.widthAnchor.constraint(equalTo: forkCustomView.widthAnchor, multiplier: 1))
+        constraints.append(sceneView.heightAnchor.constraint(equalTo: forkCustomView.heightAnchor, multiplier: 0.5))
+        constraints.append(sceneView.topAnchor.constraint(equalTo: forkCustomView.topAnchor))
+        
+        // MARK: colorButtonsStackView의 Autolayout Constraints
+        constraints.append(colorButtonsStackView.topAnchor.constraint(equalTo: styleButtonsStackView.bottomAnchor, constant: 32))
+        constraints.append(colorButtonsStackView.leadingAnchor.constraint(equalTo: forkCustomView.leadingAnchor))
+        
+        constraints.append(colorButtonsStackView.widthAnchor.constraint(equalTo: forkCustomView.widthAnchor, multiplier: 1))
+        constraints.append(colorButtonsStackView.heightAnchor.constraint(equalTo: forkCustomView.heightAnchor, multiplier: 0.2))
+        
+        // MARK: styleButtonsStackView의 Autolayout Constraints
+        constraints.append(styleButtonsStackView.topAnchor.constraint(equalTo: sceneView.bottomAnchor))
+        constraints.append(styleButtonsStackView.widthAnchor.constraint(equalTo: forkCustomView.widthAnchor, multiplier: 1))
+        constraints.append(styleButtonsStackView.heightAnchor.constraint(equalTo: forkCustomView.heightAnchor, multiplier: 0.25))
+        
+        NSLayoutConstraint.activate(constraints)
+    }
```
배열에 넣어놓고 한번에 activate 하는 방법도 있네요. 새로 하나 알아갑니다!

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        root?.childNode(withName: sample.toolToString(), recursively: true)?.isHidden = false
+    }
+    
+    // MARK: Style Button Pressed Function
```
이 부분도 코드만으로 의미 파악이 가능하므로 주석은 없어도 될 듯 합니다

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        
+        switch sender.imageView!.image {
+            
+        case UIImage(systemName: "cloud") :
```
콜론의 왼쪽 space는 지워주세용..!

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        for buttonView in view.subviews.last!.subviews {
+            buttonView.layer.borderWidth = 0
+        }
+        
```
개행을 하신 이유가 있을까요? 👀

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        let sceneInsideSceneView = SCNScene(named: "Tools.scn")
+        
+        // 모든 도구 오브젝트들을 숨깁니다.
+        
```
해당 개행은 불필요해 보입니다!

teethemoji committed:
[FIX] 회원가입뷰 배경 색상을 임시 색상 배열에서 가져오던것 수정

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        let rotateForever = SCNAction.repeatForever(action)
+        sceneInsideSceneView?.rootNode.childNode(withName: "Tools", recursively: true)?.runAction(rotateForever)
+        
+        // 렌더될 화면을 촬영할 카메라 노드를 생성합니다.
+        let cameraNode = SCNNode()
+        cameraNode.camera = SCNCamera()
+        cameraNode.camera?.fieldOfView = 60.0
+        
+        // 카메라가 있을 위치를 지정해줍니다.
+        cameraNode.position = SCNVector3(x: 0, y: 2, z: 8)
+        
+        // 카메라를 Scene 안에 배치합니다.
+        sceneInsideSceneView?.rootNode.addChildNode(cameraNode)
+        
+        // 조명을 추가합니다.
+        let lightNode = SCNNode()
+        lightNode.light = SCNLight()
+        lightNode.light?.type = .omni
+        lightNode.position = SCNVector3(x: 0, y: 10, z: 35)
+        sceneInsideSceneView?.rootNode.addChildNode(lightNode)
+        
+        // Ambient light 조명을 추가합니다.
+        let ambientLightNode = SCNNode()
+        ambientLightNode.light = SCNLight()
+        ambientLightNode.light?.type = .ambient
+        ambientLightNode.light?.color = UIColor.gray
+        sceneInsideSceneView?.rootNode.addChildNode(ambientLightNode)
```
낯선 코드라 해당 내용의 주석을 달아주셔서 의미 파악이 수월했습니다. 

필요한 `SCNNode()` 들을 외부에서 선언한 다음에 해당 UIView로 가져와 사용한다면 가독성이 더 좋아질 것 같습니다 :) 
(외부에서 선언해준 다음 `///` 주석을 통해 해당 UIView 에서 어떤 의미인지 파악할 수 있도록 해주면 어떨까요?)

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        
+        var constraints = [NSLayoutConstraint]()
+        
+        // MARK: forkCustomView의 Autolayout Constraints
+        constraints.append(forkCustomView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor))
+        constraints.append(forkCustomView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 12))
+        constraints.append(forkCustomView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -12))
+        constraints.append(forkCustomView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor))
+        
+        // MARK: sceneView의 Autolayout Constraints
+        
```
constraint 부분은 코드만으로 의미파악을 하는 것이 어렵지는 않아서, 주석은 불필요해보이는데 어떻게 생각하시나요?

lenamin commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,350 @@
+        sender.layer.borderWidth = 2
+    }
+    
+    // MARK: Color Button Pressed Function
```
이 부분 주석 또한 없더라도 의미파악에는 문제가 없어보입니다!

lenamin commented on Puhaha/Puhaha/Global/Extensioon/UIColor+Extension.swift:
```diff
@@ -37,10 +37,42 @@ extension UIColor {
+// MARK: 임의로 넣은 값이라 수정 필요
+extension UIColor {
+
+    static var customRed: UIColor {
+        return UIColor(rgb: 0xE86334)
+        // 다홍색
+    }
+    
+    static var customPurple: UIColor {
+        return UIColor(rgb: 0x7B65FF)
+        // 보라색
+    }
+    
+    static var customBlue: UIColor {
+        return UIColor(rgb: 0x6A94D1)
+        // 파란색
+    }
+    
+    static var customYellow: UIColor {
+        return UIColor(rgb: 0xFFCC1E)
+        // 노란색
+    }
+    
+    static var customBlack: UIColor {
+        return UIColor(rgb: 0x2C213A)
+        // 검은색
+    }
+    
+    static var customLightGray: UIColor {
+        return UIColor(rgb: 0xD9D9D9)
+        // 옅은 회색
+    }
```
해당 변수 위에 `///`로 색상을 적어준다면 색상을 사용할 때 옵션키를 활용해 확인할 수 있어 편리할 것 같아요.

lenamin replied:
또한 import 부분에서 `UIKit.UIColor` 만 import 해줘도 되지 않을까 합니다.

lenamin commented on Puhaha/Puhaha/Models/PokingTool.swift:
```diff
@@ -0,0 +1,62 @@
+//  Created by 김보영 on 2022/07/18.
+//
+
+import Foundation
```
UIKit이 이미 있으므로 Foundation은 불필요해보입니다. 
UIColor가 필요해 UIKit을 import 하신거라면 여기도 마찬가지로 `UIKit.UIColor`만 import 해주는 건 어떨까요?

lenamin reviewed (changes requested):
SceneView를 활용한 3D 구현 진짜 너무나 멋집니다 티모~!!!!!!! 존경스러워요!!!!!! 
불필요한 주석이나 개행만 조금 수정해주시면 가독성이 정말 좋아질 것 같습니다. 
남긴 코드리뷰 참고 부탁드립니다!

teethemoji committed:
[ADD] UIColor+Extension 속 일반 주석을 문서 주석으로 수정

- Foundation, UIKit까지 불러올 필요가 없기에 삭제하고 UIKit.UIColor만 import해오는 것으로 수정

Co-Authored-By: Lena <99120199+lenamin@users.noreply.github.com>

teethemoji committed:
Merge branch 'main' into origin/feature/teemo/temp-evacuate

teethemoji committed:
[CHORE] 불필요한 개행 & 주석 제거

teethemoji committed:
Merge branch 'main' into origin/feature/teemo/temp-evacuate

teethemoji committed:
Merge branch 'origin/feature/teemo/temp-evacuate' of https://github.com/DeveloperAcademy-POSTECH/MC3-Team3-Puhaha into origin/feature/teemo/temp-evacuate

JIW00NG commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,341 @@
+        buttonsStackView.distribution = .fillEqually
+        buttonsStackView.spacing = 16
+        
+        for i in 0..<toolImages.count {
```
```swift
for toolImage in toolImages {
```
제안 합니다.

JIW00NG commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,341 @@
+        label.font = .systemFont(ofSize: 24, weight: .semibold)
+        label.translatesAutoresizingMaskIntoConstraints = false
+        
+        let buttonsStackView = UIStackView(arrangedSubviews: [])
```
```swift
let ... = UIStackView()
```
제안 합니다

JIW00NG commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,341 @@
+        view.addSubview(forkCustomView)
+        forkCustomView.addSubview(sceneView)
+        forkCustomView.addSubview(styleButtonsStackView)
+        forkCustomView.addSubview(colorButtonsStackView)
```
```swift
[sceneView, styleButtonsStackView, colorButtonsStackView].forEach {
    forkCustomView.addSubview($0)
}
```
제안 합니다.

JIW00NG commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,341 @@
+        addConstraints()
+    }
+    
+    private func addConstraints() {
```
configureConstraints로 이름 수정해주세요

JIW00NG commented on Puhaha/Puhaha/Controllers/PokeToolViewController.swift:
```diff
@@ -0,0 +1,341 @@
+        
+        var constraints = [NSLayoutConstraint]()
+        
+        constraints.append(forkCustomView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor))
```
```swift
NSLayoutConstraint.activate([
  forkCustomView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
  ...
])
```
제한합니다.

SohyeonKim-dev commented on Puhaha/Puhaha/Global/Extensioon/UIColor+Extension.swift:
```diff
@@ -39,40 +38,41 @@ extension UIColor {
 #if DEBUG
 
-let toolColors = [UIColor.customRed, UIColor.customBlue, UIColor.customYellow, UIColor.customBlack, UIColor.customPurple]
+let toolColors = [UIColor.customLightGray, UIColor.customBlue, UIColor.customYellow, UIColor.customBlack, UIColor.customPurple]
```
오오옹 다른 색깔로 변했네요?! 😍

SohyeonKim-dev reviewed (approved):
🍴🥄🍚  티모.. 대박 ... . . 👏👏
도구들 돌아가는거 너무 멋있어요 ㅠㅡㅜㅜ
이렇게나 high-퀄리티가 되다니,, 감동의 눈물 🥲
수고하셨습니다!! :)

teethemoji committed:
[RENAME] 변수 & 함수 이름 변경

+ 네비게이션 타이틀 & 아이템 추가

teethemoji committed:
[MOVE] 백업 파일과 섞여있던 레퍼런스 위치 수정

teethemoji committed:
[REFACTOR] 배열에 append하지 않고 미리 배열 선언하는 것으로 변경

Co-Authored-By: Lena <99120199+lenamin@users.noreply.github.com>
Co-Authored-By: Ji Woong Kang <56468120+JIW00NG@users.noreply.github.com>

teethemoji committed:
[FEAT] 수평선view 추가

- 화면에 수평선을 그려주는 UIView를 return 해주는 UIHorizonDividerView 구현
- SwiftUI의 Divider()와 같은 것이라고 보면 OK

Co-Authored-By: Lena <99120199+lenamin@users.noreply.github.com>

JIW00NG reviewed (changes requested):
제안한 부분 수정부탁드려요

teethemoji committed:
[CHORE] 옅은 회색을 더 옅은 값으로 수정

teethemoji committed:
[REFACTOR] 코드 리뷰 내용 반영

- sceneView 속 환경 세팅을 위해 SCNNode 선언하고 적용하는 부분을 밖으로 빼기
- 여러 개의 view를 하나의 view의 subView로 넣을때 배열과 forEach를 이용해서 넣는 방법 적용

Co-Authored-By: Ji Woong Kang <56468120+JIW00NG@users.noreply.github.com>
Co-Authored-By: Lena <99120199+lenamin@users.noreply.github.com>

teethemoji committed:
[CHORE] 불필요한 띄어쓰기 삭제

Co-Authored-By: Lena <99120199+lenamin@users.noreply.github.com>

teethemoji committed:
[CHORE] 불필요한 주석과 argument 입력 삭제

Co-Authored-By: Lena <99120199+lenamin@users.noreply.github.com>
Co-Authored-By: Ji Woong Kang <56468120+JIW00NG@users.noreply.github.com>

lenamin commented on Puhaha/Puhaha/Global/Extensioon/UIColor+Extension.swift:
```diff
@@ -39,40 +38,42 @@ extension UIColor {
 // MARK: 임의로 넣은 값이라 수정 필요
 extension UIColor {
-
+    
+    /// 다홍색
     static var customRed: UIColor {
         return UIColor(rgb: 0xE86334)
-        // 다홍색
     }
     
```
티모 초록색 추가해주세요~!

teethemoji committed:
Merge branch 'main' into origin/feature/teemo/temp-evacuate

JIW00NG reviewed (approved):
전면 수정이 있었네요.
고생하셨어요!

lenamin reviewed (approved):
고생많으셨습니다!

teethemoji committed:
Merge branch 'main' into origin/feature/teemo/temp-evacuate
